### DISCUSSION TITLE
Reloading config function contents

### DISCUSSION THREAD

**Discussion by ehaynes99** [wbthomason/packer.nvim#1140] [General]
I've been trying to organize my setup to use config functions for setup. Aside from preventing sync issues when plugins aren't installed, it just feels like a saner strategy for managing everything than including everything in the packer config and then spreading out the config logic in other files. It should preclude any issues with things being out of sync, etc.

```lua
-- onedark.lua
return {
  'navarasu/onedark.nvim',
  commit = 'df090f9d72d43aa51dec5760c44da288b58a79b6',
  config = function()
    local onedark = require('onedark')
    onedark.setup({
      style = 'darker',
    })

    onedark.load()
  end,
}
```

I've found that the contents of `packer_compiled.lua` contain a compiled version of these functions:
```lua
  ["onedark.nvim"] = {
    config = { "\27LJ\2\nV\0\0\4\0\5\0\t6\0\0\0'\2\1\0B\0\2\0029\1\2\0005\3\3\0B\1\2\0019\1\4\0B\1\1\1K\0\1\0\tload\1\0\1\nstyle\vdarker\nsetup\fonedark\frequire\0" },
    loaded = true,
    path = "/home/erich/.local/share/nvim/site/pack/packer/start/onedark.nvim",
    url = "https://github.com/navarasu/onedark.nvim"
  },
```
However, this content is not updated by `:PackerSync` when something changes
```diff
  config = function()
     local onedark = require('onedark')
     onedark.setup({
-      style = 'darker',
+      style = 'warmer',
     })
 
     onedark.load()
```
So far, deleting `packer_compiled.lua` and then rerunning `:PackerCompile` is the only way I have found to refresh it. Does anyone have a more idiomatic solution for this? I can throw a command together to do that, but I feel like surely I'm not the first to encounter this...

**Comment by yutkat** [comment]
I write the plugin config in a separate file and load it.

```lua
-- onedark.lua
return {
  'navarasu/onedark.nvim',
  commit = 'df090f9d72d43aa51dec5760c44da288b58a79b6',
  config = function()
	require("onedark_config")
  end,
}
```


```lua
-- onedark_config.lua
local onedark = require('onedark')
onedark.setup({
  style = 'darker',
})

onedark.load()
```

**Comment by ehaynes99** [comment]
EDIT: removing this... it had other issues

**Reply by nyngwang** [reply]
@ehaynes99 Does neovim automatically run those files `lua/autocommands/*.lua` or do I have to `require` it from somewhere?

**Reply by ehaynes99** [reply]
neovim doesn't automatically run anything other than the root `init.lua`, so everything would be required there.

I should strike though the above, though. I ran into issues with loading them that way because of ordering. In general, I have just come to believe that `config` is going to be flaky regardless. It runs after a specific plugin is loaded in the middle of asynchronously loading other stuff. There are many "chicken and egg" problems that can come up, so just backed away from it altogether. Instead, I'm explicitly running configure scripts after packer is done syncing everything. `plugins/config/init.lua` explicitly requires the config scripts, which lets me explicitly set the order. How it gets required is conditional:
* if packer is installed and the compiled packer file exists on startup, it's required right away
* If not, create an autocmd that listens for `PackerComplete`, then requires it. That event is fired when `packer.sync()` is complete

Seems to be working ok. 

```lua
local ensure_packer = function()
  local fn = vim.fn
  local install_path = fn.stdpath('data') .. '/site/pack/packer/start/packer.nvim'
  if fn.empty(fn.glob(install_path)) > 0 then
    fn.system({ 'git', 'clone', '--depth', '1', 'https://github.com/wbthomason/packer.nvim', install_path })
    vim.cmd([[packadd packer.nvim]])
    return true
  end
  return false
end

local packer_bootstrap = ensure_packer()

local packer = require('packer')

local compiled_packer_exists = function()
  local f = io.open(vim.g.COMPILED_PACKER_FILE, 'r')
  if f ~= nil then
    io.close(f)
    return true
  end
end

-- remove compiled plugins when any file in `lua/plugins` changes
vim.api.nvim_create_autocmd({ 'BufWrite' }, {
  pattern = vim.g.PLUGINS_DIR .. '**',
  callback = function(args)
    local dirty = vim.fn.getbufinfo(args.buf)[1].changed ~= 0
    if dirty and compiled_packer_exists() then
      os.remove(vim.g.COMPILED_PACKER_FILE)
      print('Removed compiled plugins')
    end
  end,
})

packer.init({
  display = {
    open_fn = function()
      return require('packer.util').float({ border = 'rounded' })
    end,
  },
  git = {
    default_url_format = 'git@github.com:%s',
  },
})

return packer.startup(function(use)
  for _, plugin in pairs(require('plugins.plugins')) do
    use(plugin)
  end
  if packer_bootstrap or not compiled_packer_exists() then
    -- packer needs to sync first
    vim.api.nvim_create_autocmd({ 'User' }, {
      pattern = { 'PackerComplete' }, -- packer fires this user event
      callback = function()
        -- NOW we can configure
        require('plugins.config')
        -- deletes this autocmd. Not really necessary, but we won't need it again
        return true
      end,
    })
    require('packer').sync()
  else
    -- packer up to date, so go ahead and configure
    require('plugins.config')
  end
end)

```

**Reply by nyngwang** [reply]
If I prefer to put `require('...').setup {...}` inside `config = ...` that is provided by packer.nvim, then how should I apply your method?

**Reply by ehaynes99** [reply]
That's what I wanted as well. I just gave up on it because it didn't seem worth the effort. The dependency matrix gets way too complicated. E.g. many of the ui component plugins have to come after the colorscheme, so you have to add that dependency to all of them. The `requires` matrix just got too out of hand. Especially with the `LSP`, `cmp`, and formatters, some of them are actually cyclical dependencies that I just couldn't get to work out until they're all on disk.

Doing it this way, I can avoid the pattern I see a lot of people doing of just:
* sync and watch a bunch of explosions in the errors messages
* quit nvim
* start it again

If you're determined, the idea would be the same, however:
* nuke `packer_compiled.lua` when any config files change
* make packer sync if `packer_compiled.lua` isn't present on startup

With the above, the last bit would just change to:
```lua
return packer.startup(function(use)
  -- you could just inline the `use` statements like the examples if you prefer,
  -- but with the config scripts, it'll get huge
  for _, plugin in pairs(require('plugins.plugins')) do
    use(plugin)
  end

  if packer_bootstrap or not compiled_packer_exists() then
    require('packer').sync()
end
```
